File: GuardianMonsters/inventory/src/main/java/de/limbusdev/guardianmonsters/inventory/TeamSubMenu.java
package main.java.de.limbusdev.guardianmonsters.inventory;

import com.badlogic.gdx.scenes.scene2d.Group;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.Touchable;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.ImageButton;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable;
import com.badlogic.gdx.utils.Align;

import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian;
import de.limbusdev.guardianmonsters.guardians.monsters.Team;
import de.limbusdev.guardianmonsters.media.IMediaManager;
import de.limbusdev.guardianmonsters.services.Services;
import de.limbusdev.guardianmonsters.ui.Constant;
import main.java.de.limbusdev.guardianmonsters.inventory.ui.widgets.team.ATeamChoiceWidget;
import main.java.de.limbusdev.guardianmonsters.inventory.ui.widgets.team.MonsterStatusInventoryWidget;
import main.java.de.limbusdev.guardianmonsters.inventory.ui.widgets.team.StatusPentagonWidget;
import main.java.de.limbusdev.guardianmonsters.inventory.ui.widgets.team.TeamCircleWidget;

/**
 * Created by Georg Eckert 2017
 */

public class TeamSubMenu extends AInventorySubMenu
{

    private StatusPentagonWidget statPent;
    private MonsterStatusInventoryWidget monsterStats;
    private Image monsterImg;
    private Image blackOverlay;
    private TeamCircleWidget circleWidget;
    private ATeamChoiceWidget.Callbacks choiceHandler, swapHandler;
    private Team team;
    private ImageButton joinsBattleButton;
    private Group monsterChoice;
    private ImageButton.ImageButtonStyle lockedButtonStyle, normalButtonStyle;

    public TeamSubMenu(Skin skin, Team team) {
        super(skin);
        IMediaManager media = Services.getMedia();
        this.team = team;

        monsterChoice = new Group();
        monsterChoice.setSize(140, Constant.HEIGHT-36);
        monsterChoice.setPosition(0,0, Align.bottomLeft);
        Image monsterChoiceBg = new Image(skin.getDrawable("menu-col-bg"));
        monsterChoiceBg.setPosition(2,2,Align.bottomLeft);
        monsterChoice.addActor(monsterChoiceBg);

        choiceHandler = new ATeamChoiceWidget.Callbacks() {
            @Override
            public void onTeamMemberButton(int position) {
                showGuardianInformation(position);
            }
        };

        swapHandler = new ATeamChoiceWidget.Callbacks() {
            @Override
            public void onTeamMemberButton(int position) {
                System.out.println("Clicked " + position);
                int oldPos = circleWidget.getOldPosition();
                if(position != oldPos) {
                    AGuardian currentGuardian = team.get(oldPos);
                    AGuardian guardianToSwapWith = team.get(position);

                    team.put(oldPos, guardianToSwapWith);
                    team.put(position, currentGuardian);

                    circleWidget.init(team);
                    showGuardianInformation(position);
                }
                circleWidget.setHandler(choiceHandler);
                blackOverlay.remove();
            }
        };

        circleWidget = new TeamCircleWidget(skin, team, choiceHandler);
        circleWidget.setPosition(1,40,Align.bottomLeft);
        monsterChoice.addActor(circleWidget);

        blackOverlay = new Image(skin.getDrawable("black-a80"));
        blackOverlay.setSize(Constant.WIDTH, Constant.HEIGHT);
        ImageButton swapButton = new ImageButton(skin, "button-switch");
        swapButton.setPosition(8,8,Align.bottomLeft);
        swapButton.addListener(new com.badlogic.gdx.scenes.scene2d.utils.ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                circleWidget.remove();
                circleWidget.setHandler(swapHandler);
                addActor(blackOverlay);
                addActor(circleWidget);
            }
        });
        monsterChoice.addActor(swapButton);

        joinsBattleButton = new ImageButton(skin, "button-check");
        joinsBattleButton.setPosition(140-8,8,Align.bottomRight);
        joinsBattleButton.addListener(new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                if(joinsBattleButton.isChecked()) {
                    team.setActiveTeamSize(team.getActiveTeamSize() + 1);
                } else {
                    team.setActiveTeamSize(team.getActiveTeamSize() - 1);
                }
                System.out.println("Now active in combat: " + team.getActiveTeamSize());
            }
        });
        normalButtonStyle = joinsBattleButton.getStyle();
        lockedButtonStyle = new ImageButton.ImageButtonStyle();
        lockedButtonStyle.checked = skin.getDrawable("button-check-down-locked");

        monsterStats = new MonsterStatusInventoryWidget(skin);
        monsterStats.setPosition(140+2,0,Align.bottomLeft);
        monsterStats.init(team.get(0));


        Group monsterView = new Group();
        monsterView.setSize(140, Constant.HEIGHT-36);
        monsterView.setPosition((140+2)*2,0,Align.bottomLeft);
        Image monsterViewBg = new Image(skin.getDrawable("menu-col-bg"));
        monsterViewBg.setPosition(2,2,Align.bottomLeft);
        monsterView.addActor(monsterViewBg);
        monsterImg = new Image();
        monsterImg.setSize(128,128);
        monsterImg.setPosition(6,202,Align.topLeft);
        monsterView.addActor(monsterImg);

        statPent = new StatusPentagonWidget(skin);
        statPent.setPosition(20+2,4,Align.bottomLeft);
        monsterView.addActor(statPent);

        addActor(monsterChoice);
        addActor(monsterStats);
        addActor(monsterView);

        showGuardianInformation(0);

        setDebug(Constant.DEBUGGING_ON, true);
    }

    private void showGuardianInformation(int teamPosition) {
        monsterStats.init(team.get(teamPosition));
        monsterImg.setDrawable(new TextureRegionDrawable(
            Services.getMedia().getMonsterSprite(team.get(teamPosition).getSpeciesDescription().getID(), team.get(teamPosition).getAbilityGraph().getCurrentForm())));
        statPent.init(team.get(teamPosition));
        joinsBattleButton.remove();
        joinsBattleButton.setChecked(false);
        joinsBattleButton.setStyle(normalButtonStyle);

        // If the shown position belongs to the range given by activeInCombat & within 0-2
        if(teamPosition <= team.getActiveTeamSize() && teamPosition <3 && teamPosition < 3) {
            // TODO take max team size into account
            joinsBattleButton.setTouchable(Touchable.enabled);
            monsterChoice.addActor(joinsBattleButton);

            // if shown monster is in the range of active monsters
            if(teamPosition < team.getActiveTeamSize()) {
                joinsBattleButton.setChecked(true);
            }
            // the shown monster is at position 0 or not the last active monster
            if(teamPosition == 0 || teamPosition < team.getActiveTeamSize() -1) {
                joinsBattleButton.setTouchable(Touchable.disabled);
                joinsBattleButton.setStyle(lockedButtonStyle);
            }
        }
    }

    @Override
    public void refresh() {
        showGuardianInformation(circleWidget.getCurrentPosition());
    }
}

File: GuardianMonsters/inventory/src/de/limbusdev/guardianmonsters/inventory/TeamSubMenu.kt
package de.limbusdev.guardianmonsters.inventory

import com.badlogic.gdx.scenes.scene2d.Group
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.ImageButton
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.utils.Align

import java.util.Observable
import java.util.Observer

import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian
import de.limbusdev.guardianmonsters.guardians.monsters.Team
import de.limbusdev.guardianmonsters.inventory.de.limbusdev.guardianmonsters.inventory.ui.widgets.team.BattleParticipationToggleButton
import de.limbusdev.guardianmonsters.inventory.ui.widgets.team.StatusStarGlyphWidget
import de.limbusdev.guardianmonsters.scene2d.*
import de.limbusdev.guardianmonsters.services.Services
import de.limbusdev.guardianmonsters.ui.Constant
import de.limbusdev.guardianmonsters.ui.widgets.GuardianStatusWidget
import de.limbusdev.guardianmonsters.ui.widgets.TeamCircleWidget
import de.limbusdev.utils.logDebug
import de.limbusdev.utils.logInfo
import ktx.style.get

/**
 * TeamSubMenu shows his team to the player. It allows organizing team order and which positions
 * join the fight. It shows Guardian details, like their [Equipment] and individual statistics.
 *
 * @author Georg Eckert 2017
 */

class TeamSubMenu
(
        skin: Skin = Services.UI().inventorySkin,
        private val team: Team
)
    : AInventorySubMenu(), Observer
{
    // --------------------------------------------------------------------------------------------- PROPERTIES
    companion object { const val TAG = "TeamSubMenu" }

    private lateinit var starGlyph          : StatusStarGlyphWidget
    private lateinit var monsterStats       : GuardianStatusWidget
    private lateinit var circleWidget       : TeamCircleWidget

    private lateinit var monsterImg         : Image
    private lateinit var blackOverlay       : Image

    private lateinit var joinToggleButton   : BattleParticipationToggleButton // activates a team position for battle
    private lateinit var swapButton         : ImageButton   // activates swap mode
    private lateinit var lockedStyle        : ImageButton.ImageButtonStyle
    private lateinit var defaultStyle       : ImageButton.ImageButtonStyle

    private lateinit var monsterChoice      : Group

    // ................................................................................... Callbacks
    private lateinit var onCircleWidgetChoice: (Int) -> Unit // what happens, when circle is touched
    private lateinit var onCircleWidgetSwap  : (Int) -> Unit // what happens, when circle is touched in swap mode


    // --------------------------------------------------------------------------------------------- CONSTRUCTORS
    init
    {
        layout(skin)
        setupCallbacks()

        circleWidget.initialize(team)
        monsterStats.initialize(team[0])
        showGuardianInformation(0)

        setDebug(InventoryDebugger.SCENE2D_DEBUG, true)
    }


    // --------------------------------------------------------------------------------------------- LAYOUT
    override fun layout(skin: Skin)
    {
        circleWidget = TeamCircleWidget(skin)

        monsterChoice = makeGroup(width = 140f, height = Constant.HEIGHTf - 36)

        val monsterChoiceBg = makeImage(skin["menu-col-bg"], ImgPosition(2f, 2f), monsterChoice)

        circleWidget.setPosition(1f, 40f, Align.bottomLeft)
        monsterChoice.addActor(circleWidget)

        blackOverlay = makeImage(skin["black-a80"], ImgLayout(Constant.WIDTHf, Constant.HEIGHTf))
        swapButton = makeImageButton(skin["button-switch"], PositionXYA(8f, 8f), monsterChoice)

        joinToggleButton = BattleParticipationToggleButton(skin, PositionXYA(140-8f, 8f, Align.bottomRight))

        defaultStyle = joinToggleButton.style
        lockedStyle = ImageButton.ImageButtonStyle().apply { checked = skin["button-check-down-locked"] }

        val monsterView = makeGroup(140f, Constant.HEIGHTf - 36, (140+2)*2f, 0f)
        val monsterViewBg = makeImage(skin["menu-col-bg"], ImgPosition(2f, 2f), monsterView)
        monsterImg = Image()
        monsterImg.setup(ImgLayout(128f, 128f, 6f, 202f, Align.topLeft), monsterView)

        monsterStats = GuardianStatusWidget()
        monsterStats.setPosition(140f + 2, 0f, Align.bottomLeft)

        starGlyph = StatusStarGlyphWidget(skin)
        starGlyph.setPosition(20f + 2, 4f, Align.bottomLeft)
        monsterView.addActor(starGlyph)

        addActor(monsterChoice)
        addActor(monsterStats)
        addActor(monsterView)

        updateJoinBattleHighlights()
    }

    private fun showGuardianInformation(teamPosition: Int)
    {
        logDebug(TAG) { "$teamPosition" } // is called twice, due to propagateSelectedGuardian(...)
        val guardian = team[teamPosition]

        // Update Stats Widget, Image and Stats Pentagram
        monsterStats.initialize(guardian)
        monsterImg.setRegion(Services.Media().getMonsterSprite(guardian.speciesID, guardian.currentForm))
        starGlyph.initialize(guardian)

        // Reset Join Battle Button
        joinToggleButton.remove()

        // Only team positions 0..2 can be activated
        joinToggleButton.autoEnableOrDisable(teamPosition, team.activeTeamSize, monsterChoice)
    }


    // --------------------------------------------------------------------------------------------- SYNCHRONISATION
    override fun refresh() {}

    /** Guardian which are activated to join the battle, are highlighted with a red circle. */
    private fun updateJoinBattleHighlights()
    {
        for(i in 0..6) { circleWidget.unhighlight(i) }
        for(i in 0 until team.activeTeamSize) { circleWidget.highlight(i) }
    }

    override fun syncSelectedGuardian(teamPosition: Int)
    {
        // Update widgets
        circleWidget.initialize(team, teamPosition)
        showGuardianInformation(teamPosition)

        team.values().forEach { it.deleteObserver(this) }

        team[teamPosition].addObserver(this)
    }

    override fun update(observable: Observable, o: Any?)
    {
        if (observable is AGuardian)
        {
            showGuardianInformation(team.getPosition(observable))
        }
    }


    // --------------------------------------------------------------------------------------------- CALLBACKS
    private fun setupCallbacks()
    {
        onCircleWidgetChoice = { position ->

            showGuardianInformation(position)
            propagateSelectedGuardian(position) // propagation calls showGuardianInformation(...)
        }

        onCircleWidgetSwap = { position ->

            if (circleWidget.oldPosition != position)
            {
                val currentGuardian    = team[circleWidget.oldPosition]
                val guardianToSwapWith = team[position]

                team.swap(circleWidget.oldPosition, position)

                circleWidget.initialize(team)
                showGuardianInformation(position)
                propagateSelectedGuardian(position)
            }

            circleWidget.setHandler(onCircleWidgetChoice)
            blackOverlay.remove()
        }

        swapButton.replaceOnClick {

            circleWidget.remove()
            circleWidget.setHandler(onCircleWidgetSwap)
            addActor(blackOverlay)
            addActor(circleWidget)
        }

        joinToggleButton.replaceOnClick {

            joinToggleButton.toggle()
            team.activeTeamSize = when(joinToggleButton.isChecked)
            {
                true  -> team.activeTeamSize + 1
                false -> team.activeTeamSize - 1
            }
            logInfo(TAG) { "Now active in combat: ${team.activeTeamSize}" }
            updateJoinBattleHighlights()
        }

        circleWidget.setHandler(onCircleWidgetChoice)
    }
}
